pkg/kubelet/kubelet.go

```diff
@@ -3070,3 +3051,83 @@ func (kl *Kubelet) PrepareDynamicResources(ctx context.Context, pod *v1.Pod) err
 func (kl *Kubelet) UnprepareDynamicResources(ctx context.Context, pod *v1.Pod) error {
 	return kl.containerManager.UnprepareDynamicResources(ctx, pod)
 }
+
+// staticPodRegistration ensures that all static pods are registered to the apiserver
+// before node became ready.
+func (kl *Kubelet) initialStaticPodsRegistration() error {
+	// kubelet running in standalone mode does not register static pods to the apiserver.
+	if kl.kubeClient == nil {
+		return nil
+	}
+	// Check if initial static pod registration has already been completed.
+	if kl.initialStaticPodsRegistered.Load() {
+		return nil
+	}
+	staticPodToMirrorPodMap := kl.podManager.GetStaticPodToMirrorPodMap()
+	for _, mirrorPod := range staticPodToMirrorPodMap {
+		if mirrorPod == nil {
+			return fmt.Errorf("not all static pods are registered")
+		}
+	}
+
+	kl.initialStaticPodsRegistered.Store(true)
+	return nil
+}
+
+// Ensure Mirror Pod for Static Pod exists.
+func (kl *Kubelet) ensureMirrorPodExists(staticPod, mirrorPod *v1.Pod) {
+	if !kubetypes.IsStaticPod(staticPod) {
+		return
+	}
+	deleted := false
+	if mirrorPod != nil {
+		if mirrorPod.DeletionTimestamp != nil || !kubepod.IsMirrorPodOf(mirrorPod, staticPod) {
+			// The mirror pod is semantically different from the static pod. Remove
+			// it. The mirror pod will get recreated later.
+			klog.InfoS("Trying to delete pod", "pod", klog.KObj(staticPod), "podUID", mirrorPod.ObjectMeta.UID)
```

Comment:
it is strange that we are logging `staticPod` in this error message, and `mirrorPod` later. I cannot think of a problem it would cause, but inconsistency looks bad